lib/mix/lib/mix/scm/git.ex

```diff
@@ -126,17 +126,16 @@ defmodule Mix.SCM.Git do
     update_origin(opts[:git])
 
     # Fetch external data
-    branch_or_tag = opts[:branch] || opts[:tag]
+    rev = get_lock_rev(opts[:lock], opts) || get_opts_rev(opts)
 
     ["--git-dir=.git", "fetch", "--force", "--quiet"]
     |> Kernel.++(progress_switch(git_version()))
     |> Kernel.++(tags_switch(opts[:tag]))
-    |> Kernel.++(depth_switch(opts[:depth]))
-    |> Kernel.++(if branch_or_tag, do: ["origin", branch_or_tag], else: [])
```

Comment:
Didn't we change it to always add ["origin", rev] on purpose so we fetch only the branch/tag we are going to use instead of all of it? So it would be nice to leave this in regardless if the depth option is being used.

Reply:
I would prefer to fetch less by default, so I think we should keep the new behaviour of only fetching branch/tag by default. Most people won't realistically add the depth option.

Reply:
That's right. I think I found a safe combination after we validate the depth+ref case:

```diff
diff --git lib/mix/lib/mix/scm/git.ex lib/mix/lib/mix/scm/git.ex
index 442c241f5..9c9bd9eea 100644
--- lib/mix/lib/mix/scm/git.ex
+++ lib/mix/lib/mix/scm/git.ex
@@ -131,7 +131,8 @@ defp checkout(_path, opts) do
     ["--git-dir=.git", "fetch", "--force", "--quiet"]
     |> Kernel.++(progress_switch(git_version()))
     |> Kernel.++(tags_switch(opts[:tag]))
-    |> Kernel.++(depth_switch(opts[:depth], rev))
+    |> Kernel.++(depth_switch(opts[:depth]))
+    |> Kernel.++(refspec_switch(opts, rev))
     |> git!()
 
     # Migrate the Git repo
@@ -211,11 +212,21 @@ defp progress_switch(version) do
   defp tags_switch(nil), do: []
   defp tags_switch(_), do: ["--tags"]
 
-  defp depth_switch(nil, _rev), do: []
+  defp depth_switch(nil), do: []
 
-  defp depth_switch(n, rev) when is_integer(n) and n > 0 do
+  defp depth_switch(n) when is_integer(n) and n > 0 do
     check_depth_support(git_version())
-    ["--depth=#{n}"] ++ if rev, do: ["origin", rev], else: []
+    ["--depth=#{n}"]
+  end
+
+  defp refspec_switch(opts, rev) do
+    case Keyword.take(opts, [:depth, :branch, :tag]) do
+      [_ | _] ->
+        ["origin", rev]
+
+      _ ->
+        []
+    end
   end
 
   ## Helpers
```

The intention is we'll fetch only a single tree if a `:tag` or a `:branch` is specified, or if `:depth` is used.

The `:ref` option is not taken into account to support short hashes.

If we want to go further in this direction, we could also try to fetch only a specific tree if `byte_size(ref) >= 40`. What do you think?